node/src/actors/chain_manager/superblock.rs

```diff
@@ -0,0 +1,331 @@
+            }
+        }
+    }
+
+    /// Since we do not check signatures here, a superblock vote is valid if the signing identity
+    /// is in the ARS.
+    /// Returns true, false, or unknown
+    fn is_valid(&self, sbv: &SuperBlockVote) -> Option<bool> {
+        match self.current_superblock_index {
+            // We do not know the current index, we cannot know if the vote is valid
+            None => None,
+            // If the index is the same as the current one, the vote is valid if it is signed by a
+            // member of the ARS
+            Some(x) if x == sbv.superblock_index => self
+                .previous_ars_identities
+                .as_ref()
+                .map(|x| x.contains(&sbv.secp256k1_signature.public_key.pkh())),
+            // If the index is not the same as the current one, but it is within an acceptable range
+            // of [x-1, x+1], broadcast the vote without checking if it is a member of the ARS, as
+            // the ARS may have changed and we do not keep older copies of the ARS in memory
+            Some(x) => {
+                // Check [x-1, x+1] range with overflow prevention
+                if ((x.saturating_sub(1))..=(x.saturating_add(1))).contains(&sbv.superblock_index) {
+                    None
+                } else {
+                    Some(false)
+                }
+            }
+        }
+    }
+
+    /// Produces a `SuperBlock` that includes the blocks in `block_headers` if there is at least one of them.
+    /// `sorted_ars_identities` will be used to validate all the superblock votes received for the
+    /// next superblock. The votes for the current superblock must be validated using
+    /// `previous_ars_identities`.
+    pub fn build_superblock(
+        &mut self,
+        block_headers: &[BlockHeader],
+        sorted_ars_identities: &[PublicKeyHash],
+        superblock_index: u32,
+        last_block_in_previous_superblock: Hash,
+    ) -> Option<SuperBlock> {
+        self.current_superblock_index = Some(superblock_index);
+        self.votes_on_local_superlock.clear();
+
+        match mining_build_superblock(
+            block_headers,
+            sorted_ars_identities,
+            superblock_index,
+            last_block_in_previous_superblock,
+        ) {
+            None => {
+                // Clear state when there is no superblock
+                // Note that the ARS members list is not updated in this case
+                self.current_superblock_hash = None;
+                self.received_superblocks.clear();
+
+                None
+            }
+            Some(superblock) => {
+                let superblock_hash = superblock.hash();
+                self.current_superblock_hash = Some(superblock_hash);
+
+                // Save ARS identities:
+                // previous = current
+                // current = sorted_ars_identities
+                {
+                    std::mem::swap(
```

Comment:
is there a test that check this hashset swap?

Reply:
Working on it

Reply:
Done, it's superblock_state_ars_identities